lib/propolis/src/hw/qemu/fwcfg.rs

```diff
@@ -396,259 +304,353 @@ impl FwCfg {
+        // selector when the SELECT function is specified.
+        if FwCfgDmaCtrl::from_bits(req.ctrl.get() & 0xffff).is_none() {
+            dma_write_result(false, GuestAddr(addr), mem);
+            return Err(FwCfgErr::UnrecognizedDmaCmd);
+        }
 
-            let mut ro = ReadOp::from_mapping(offset as usize, mapping);
-            self.xfer(selector, RWOp::Read(&mut ro))?;
-            to_write
-        } else {
-            0
-        };
+        let res = self.dma_operation(state, req, mem);
 
-        // write zeroes for everything past the end of the data
-        if written < len {
-            mem.write_byte(
-                GuestAddr(addr + u64::from(written)),
-                0,
-                (len - written) as usize,
-            );
-        }
+        dma_write_result(res.is_ok(), GuestAddr(addr), mem);
         Ok(())
     }
-    fn dma_write(
+
+    fn dma_operation(
         &self,
-        selector: u16,
-        addr: u64,
-        offset: u32,
-        len: u32,
+        state: &mut MutexGuard<State>,
+        req: FwCfgDmaAccess,
         mem: &MemCtx,
-    ) -> Result {
-        let valid_remain = self.size(selector).saturating_sub(offset);
+    ) -> Result<(), FwCfgErr> {
+        let opts = FwCfgDmaCtrl::from_bits_truncate(req.ctrl.get());
+        if opts.contains(FwCfgDmaCtrl::SELECT) {
+            let selector = (req.ctrl.get() >> 16) as u16;
+            self.select(state, selector);
+        }
 
-        if valid_remain > 0 {
-            let to_read = len.min(valid_remain);
-            let mapping = mem
+        if opts.contains(FwCfgDmaCtrl::READ) {
+            let buf_len = req.len.get() as usize;
+            let map = mem
+                .writable_region(&GuestRegion(
+                    GuestAddr(req.addr.get()),
+                    buf_len,
+                ))
+                .ok_or(FwCfgErr::BadAddr)?;
+            let mut ro = ReadOp::from_mapping(0, map);
+
+            let nread = self.read(state, &mut ro)?;
+            if nread < buf_len {
+                // If the item being read did not cover the entire DMA region in
+                // the request, zero out the rest
+                assert!(ro.avail() > 0);
+                ro.fill(0);
+            }
+        } else if opts.contains(FwCfgDmaCtrl::WRITE) {
+            let buf_len = req.len.get() as usize;
+            let map = mem
                 .readable_region(&GuestRegion(
-                    GuestAddr(addr),
-                    to_read as usize,
+                    GuestAddr(req.addr.get()),
+                    buf_len,
                 ))
-                .ok_or("bad GPA")?;
-            let mut wo = WriteOp::from_mapping(offset as usize, mapping);
-            self.xfer(selector, RWOp::Write(&mut wo))?;
+                .ok_or(FwCfgErr::BadAddr)?;
+            let mut wo = WriteOp::from_mapping(0, map);
+            self.write(state, &mut wo)?;
+        } else if opts.contains(FwCfgDmaCtrl::SKIP) {
```

Comment:
The fact that we will do a read operation even if the `WRITE` or `SKIP` bit is set, and do a write regardless of whether the `SKIP` bit is set, felt a bit sketchy to me --- until I looked at [the QEMU documentation](https://www.qemu.org/docs/master/specs/fw_cfg.html#guest-side-dma-interface) and saw that it describes this exact behavior. Perhaps we ought to have a comment explaining that?

Reply:
Yeah a comment is a good suggestion.  Hopefully the addition is suitable?

Reply:
Looks good to me!